File: automation/jsr223/python/personal/heating_do_heaters_need_changing.py
from core.rules import rule
from core.triggers import when
from core.actions import LogAction
from core.actions import ScriptExecution
from java.time import ZonedDateTime as DateTime

@rule("Check if Heaters need changing etc", description="Check if Heaters need changing etc", tags=["heating"])
@when("Member of gHeatingModes received update")
@when("Member of gTemperatureSetpoints received update")
@when("Member of gRoomTemperatures received update")
def checkIfHeatersNeedUpdating(event):
    # checkIfHeatersNeedUpdating.log.info("HHH HeatingMode, Setpoint or Temperature updated")
    LogAction.logInfo("Check if Heaters need changing", "HHH HeatingMode, Setpoint or Temperature updated")

    # get prefix eg FR, CT etc
    prefix = event.itemName[:event.itemName.rfind('_')]
    LogAction.logDebug("Check if Heaters need changing", "HHH Check if Heaters need changing etc due to Item: {}, received  update: {}", event.itemName, event.itemState)

    HeatingMode = ir.getItem(prefix + "_HeatingMode")
    LogAction.logDebug("Check if Heaters need changing", "HHH HeatingMode {} : {}", prefix, HeatingMode.state)

    TSetpoint = ir.getItem(prefix + "_TemperatureSetpoint")
    LogAction.logDebug("Check if Heaters need changing", "HHH Setpoint    {} : {}", prefix, TSetpoint.state)

    Temperature = ir.getItem(prefix + "_Temperature")
    LogAction.logDebug("Check if Heaters need changing", "HHH Temperature {} : {}", prefix, Temperature.state)

    Heater = ir.getItem(prefix + "_Heater")
    LogAction.logDebug("Check if Heaters need changing", "HHH Heater      {} : {}", prefix, Heater.state)

    Reachable = ir.getItem(prefix + "_RTVReachable")
    LogAction.logDebug("Check if Heaters need changing", "HHH Reachable   {} : {}", prefix, Reachable.state)

    #!handle an offline TRV
    if Reachable.state.toString() != "Online":  # is the trv actually online??
       LogAction.logDebug("Check if Heaters need changing", "HHH ZZZZ---ZZZZ Reachable-Offline - sending OFF, leaving!!!!!   prefix: {} ", prefix)
       #turn this one off
       events.sendCommand(Heater, "OFF")
       return #dont continue on and update the bolier control if this RTV is Offline

    LogAction.logDebug("Check if Heaters need changing", "HHH current HeatingMode.state: {}  masterHeatingMode.state: {}",
                       HeatingMode.state, ir.getItem("masterHeatingMode").state)

    if (HeatingMode.state.toString() == "off") or (ir.getItem("masterHeatingMode").state.toString() == "off"):
        if (ir.getItem("masterHeatingMode").state.toString() == "off"):
            LogAction.logDebug("Check if Heaters need changing", "HHH Master Heating Mode is OFF")

        LogAction.logDebug("Check if Heaters need changing", "HHH Turn heater OFF for {}  cos its Heating Mode is {}", prefix, HeatingMode.state)
        events.sendCommand(Heater, "OFF")

    #* if alowwed to be on, check if need to turn on heater
    elif (HeatingMode.state.toString() == "auto") or (HeatingMode.state.toString() == "manual"):
        LogAction.logDebug("Check if Heaters need changing", "HHH mode is auto or manual")
        LogAction.logDebug("Check if Heaters need changing", "HHH Heater.itemName: {}", Heater)
        checkIfHeatersNeedUpdating.log.debug(":::-> Check if Heaters need changing  HHH mode is auto or manual")
        setpoint = TSetpoint.state
        turnOnTemp = setpoint  # - 0.2// calculate the turn on/off temperatures
        turnOffTemp = setpoint  # + 0.1
        temp = Temperature.state  # get the current temperature

        if temp >= turnOffTemp:  # {  // determine whether we need to turn on/off the heater
            LogAction.logDebug("Check if Heaters need changing", "HHH SendCommand to {}, Heater OFF", prefix)
            events.sendCommand(Heater, "OFF")
        elif temp < turnOnTemp:
            LogAction.logDebug("Check if Heaters need changing", "HHH SendCommand to {}, Heater ON", prefix)
            events.sendCommand(Heater, "ON")

File: automation/js/heating_do_heaters_need_changing.js
const {
  log, items, rules, actions, time, triggers,
} = require('openhab');
const { myutils } = require('personal');

const logger = log('heater change?');
const { countdownTimer, timeUtils, timerMgr } = require('openhab_rules_tools');

// if gHeatingModes, gTemperatureSetpoints ,gRoomTemperatures are updated
// figure out if a room heater needs turning on
rules.JSRule({
  name: 'Check if Heaters need changing etc',
  description: 'Check if Heaters need changing etc',
  triggers: [
    triggers.GroupStateUpdateTrigger('gHeatingModes'),
    triggers.GroupStateUpdateTrigger('gTemperatureSetpoints'),
    triggers.GroupStateUpdateTrigger('gRoomTemperatures'),
    // triggers.GroupStateChangeTrigger('gHeaterBoosters', 'OFF', 'ON'), // on edges only
    // triggers.GroupStateChangeTrigger('gHeaterBoosters', 'ON', 'OFF'),
  ],
  execute: (event) => {
    // console.log(event);
    // logger.warn('>Mode, setpoint or temp changed. Do any Heaters need . changing etc?');
    // const action = 'default';
    // get prefix eg FR, CT etc
    const roomPrefix = event.itemName.toString().substr(0, event.itemName.lastIndexOf('_'));

    const heatingModeItem = items.getItem(`${roomPrefix}_HeatingMode`);
    // logger.warn(`>heatingModeItem.name: ${heatingModeItem.name} : ,  heatingModeItem.state: ${heatingModeItem.state}`);

    const setpointItem = items.getItem(`${roomPrefix}_TemperatureSetpoint`);
    // logger.warn(`>setpointItem.name: ${setpointItem.name} : ,  Setpoint.state: ${setpointItem.state}`);

    const TemperatureItem = items.getItem(`${roomPrefix}_Temperature`);
    // logger.warn(`>TemperatureItem.name: ${TemperatureItem.name} : ,  TemperatureItem.state: ${TemperatureItem.state}`);

    const HeaterItem = items.getItem(`${roomPrefix}_Heater`);
    // logger.warn(`>HeaterItem.name: ${HeaterItem.name} : ,  HeaterItem.state: ${HeaterItem.state}`);

    const ReachableItem = items.getItem(`${roomPrefix}_RTVReachable`);
    // logger.warn(`>ReachableItem.name: ${ReachableItem.name} : ,  ReachableItem.state: ${ReachableItem.state}`);

    // !handle an offline TRV - return
    // !if ANY trvs are unreachable - turn off hetarer to prevent false demand
    // not just the calkkling device - which cant call anyway as its offline
    // dont continue on and update the bolier control if this RTV is Offline
    if (ReachableItem.state.toString() !== 'Online') {
      logger.warn(`>>ZZZZ ReachableItem-Offline - sending OFF, leaving!!!!! : ${roomPrefix} : ,  ReachableItem.state: ${ReachableItem.state}`);
      // turn it off
      HeaterItem.sendCommand('OFF');

      items.getItem('HL_Heater').sendCommand('OFF');
      // dont continue on and update the bolier control if this RTV is Offline
      return;
    }
    items.getItem('HL_Heater').sendCommand('OFF');

    // logger.warn(`>masterHeatingMode.state.toString() : ${items.getItem('masterHeatingMode').state.toString()}`);

    //! add if boost on - skip
    // if this heater is currently in being boosted, then just l;eave it alone and move on
    const BoostItem = items.getItem(`${roomPrefix}_Boost`, true);// get boost item for this heater, return null if missing
    if (BoostItem && BoostItem.state.toString() === 'ON') {
      logger.error(`...........................>>>>Boosting item -->> BoostItem.name, return from heater routine: ${BoostItem.name}`);
      if (BoostItem.state === 'ON') { // in a boost period
        return;
      }
    }
    logger.debug('>>>>no boost item defined for this heater, process as normal');
    // if HEATER alowed to be on, check if need to turn on heater
    if (((heatingModeItem.state.toString() === 'auto')) || ((heatingModeItem.state.toString() === 'manual'))) {
      // logger.warn(`>>Heater: ${roomPrefix}, mode is: ${heatingModeItem.state.toString()}`);
      const setpoint = setpointItem.rawState;
      const turnOnTemp = setpoint; // # - 0.2// calculate the turn on/off temperatures
      const turnOffTemp = setpoint; //  # + 0.1
      const temp = TemperatureItem.rawState; //  # get the current temperature
      if (temp >= turnOffTemp) {
        logger.warn(`Heating change... Heater: ${roomPrefix}, mode is: ${heatingModeItem.state.toString()} -> SendCommand to: ${roomPrefix}, Heater OFF`);
        HeaterItem.sendCommand('OFF');
      } else if (temp < turnOnTemp) {
        logger.warn(`Heating change... Heater: ${roomPrefix}, mode is: ${heatingModeItem.state.toString()} -> SendCommand to: ${roomPrefix}, Heater ON`);
        HeaterItem.sendCommand('ON');
      }
    } else if ((heatingModeItem.state.toString() === 'off') || (items.getItem('masterHeatingMode').state.toString() === 'off')) {
      if ((items.getItem('masterHeatingMode').state.toString() === 'off')) {
        logger.warn('Heating change... Master Heating Mode is OFF!');
      }
      logger.warn(`Heating change...Turn ${roomPrefix} heater OFF, :. its Heating Mode is  ${heatingModeItem.state}`);
      HeaterItem.sendCommand('OFF');
    }
  },
});
